GitUI/BranchTreePanel/RepoObjectsTree.cs

```diff
@@ -163,34 +164,44 @@ public async Task ReloadAsync()
                 treeMain.BeginUpdate();
                 _rootNodes.ForEach(t => t.IgnoreSelectionChangedEvent = true);
                 var tasks = _rootNodes.Select(r => r.ReloadAsync(token)).ToArray();
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                await Task.WhenAll(tasks).ConfigureAwait(true);
             }
             finally
             {
-                await this.SwitchToMainThreadAsync();
+                ThreadHelper.AssertOnUIThread();
+
                 if (!token.IsCancellationRequested)
                 {
                     Enabled = true;
+                }
+
+                // Need to end the update for the selected node to scroll into view below under certain conditions
+                // (e.g. when the treeview gets larger, trying to make the selected node visible doesn't work).
+                treeMain.EndUpdate();
+
+                if (!token.IsCancellationRequested)
+                {
                     var selectedNode = treeMain.AllNodes().FirstOrDefault(n =>
                         _rootNodes.Any(rn => $"{rn.TreeViewNode.FullPath}{treeMain.PathSeparator}{currentBranch}" == n.FullPath));
                     if (selectedNode != null)
                     {
-                        treeMain.SelectedNode = selectedNode;
-                        ExpandPathToSelectedNode();
+                        SetSelectedNode(selectedNode);
                     }
 
                     _rootNodes.ForEach(t => t.IgnoreSelectionChangedEvent = false);
                 }
-
-                treeMain.EndUpdate();
             }
+        }
 
-            void ExpandPathToSelectedNode()
-            {
-                treeMain.Scrollable = false; // disable scrolling, so the next call does not horizontally (nor vertically) scroll
-                treeMain.SelectedNode.EnsureVisible();
-                treeMain.Scrollable = true;
-            }
+        private void SetSelectedNode(TreeNode node)
+        {
+            treeMain.SelectedNode = node;
+            treeMain.SelectedNode.EnsureVisible();
+
+            // EnsureVisible leads to horizontal scrolling in some cases. We make sure to force horizontal
+            // scroll back to 0. Note that we use SendMessage rather than SetScrollPos as the former works
+            // outside of Begin/EndUpdate.
+            NativeMethods.SendMessageInt((IntPtr)treeMain.Handle, NativeMethods.WM_HSCROLL, (IntPtr)NativeMethods.SB_LEFT, IntPtr.Zero);
```

Comment:
NativeMethods had to check for RunningOnWindows() when there was Mono compatibility. I believe this is OK right now, if we go multi platform again, this is easy enough to identify.